Fix Client Command Dispatcher Thread hijacking

tkirill opened the pull request:
Fix #623

Replace Client Command Dispatched Thread with critical section around `persistentChannel`.  I'm afraid that there was a special reason why the Client Command Dispatched Thread was implemented instead of critical section which I can't understand.

zidad commented:
Great thanks a lot! Did you run all the unit tests? There seems to be a failing one but I can't see which one in AppVeyor. 

Would you mind increasing the version number in version.cs?

micdenny commented:
I need to read carefully the #623

tkirill commented:
@zidad The failed test was checking that action passed to `ClientCommandDispatcher.InvokeAsync` is invoked on Client Command Dispatcher Thread.

I fixed a test and add a few other tests.  I rebased my commits on the latest master, sorry for that.  Now the build is green and I will not rebase during the review.

@micdenny Yes, this is a pretty radical change.  Please don't think I want to rush in this really good project and ruin it by refactoring the core from the start :)  I just want to provide a starting point for discussion about the problem and possible ways to fix it.

micdenny commented:
Thank you @tkirill in fact I would like to deeply investigate the related issue, it is well written, so it completely took my attention :)

zidad commented:
@micdenny I've tested this change and it looks pretty stable for me. We could also consider just merging it to our feature/netcore branch as that needs to be fully tested for stability anyway?

ebbeflarup commented:
Hello
We're experiencing this exact problem. Is there any news on when this pull request will be merged? 

Thank in advance :)

zidad commented:
@ebbeflarup done!

micdenny commented:
@tkirill @zidad probably we have a problem with this implementation check #661
